### PR TITLE
refactor(core): update with-routes to handle redirect logic better

### PR DESCRIPTION
**Pull request by chanceaclark**
## What/Why?
Merging 3rd party code from #1429 (PR was created to run CI)

## Testing
See CI

### CODE REVIEW

**Review by deini — APPROVED** [review]

File: core/middlewares/with-routes.ts [hunk 1]
```diff
@@ -251,20 +269,21 @@ export const withRoutes: MiddlewareFactory = () => {
...
-          return NextResponse.redirect(route.redirect.toUrl, redirectConfig);
+        case 'BlogPostRedirect':
+        case 'BrandRedirect':
+        case 'CategoryRedirect':
+        case 'PageRedirect':
+        case 'ProductRedirect': {
+          // For dynamic redirects, assume an internal redirect and construct the URL from the path
+          const redirectUrl = new URL(route.redirect.to.path, request.url);
+
+          return NextResponse.redirect(redirectUrl, redirectConfig);
```
Comment (chanceaclark) [inline core/middlewares/with-routes.ts:280]: Hey @willPrattUPL I found a bug in the previous logic as it was rewriting, not redirecting. Can you test y'all redirects from this branch, specifically that 404 that had `node` return `null`?

Comment (chanceaclark) [reply core/middlewares/with-routes.ts:286]: I manually tested this with your 404 response, but added a `path` node to the "response". It seemed to work on my end but had to push a change to make sure trailing slashes were preserved correctly.

Comment (willPrattUPL) [reply core/middlewares/with-routes.ts:286]: Hey, sorry Chance. We are extremely busy right now but I promise I'll get to this this week @chanceaclark

**Review by deini — APPROVED** [review]

**Comment by willPrattUPL** [comment]
@chanceaclark 

I've run this twice, both on the latest changes of my forked repo and our >2months old fork that we have been building off of

In both cases I get this:

image

on all redirects

https://github.com/bigcommerce/catalyst/pull/1429#discussion_r1793476393

/uplift-v2-standing-desk-v2-or-v2-commercial/
/uplift-v2-l-shaped-standing-desk/
/uplift-v2-standing-desk-frame/
/height-adjustable-standing-desk-with-l-shaped-top/



All redirects exhibit the exact same behavior as before, except now the 404 is an infinite redirect loop

[
@bigcommerce/catalyst-core:dev:   {
@bigcommerce/catalyst-core:dev:     "route": {
@bigcommerce/catalyst-core:dev:       "redirect": {
@bigcommerce/catalyst-core:dev:         "to": {
@bigcommerce/catalyst-core:dev:           "__typename": "ProductRedirect",
@bigcommerce/catalyst-core:dev:           "path": "/uplift-v2-l-shaped-standing-desk/"
@bigcommerce/catalyst-core:dev:         },
@bigcommerce/catalyst-core:dev:         "toUrl": "https://www.upliftdesk.com/uplift-v2-l-shaped-standing-desk/"
@bigcommerce/catalyst-core:dev:       },
@bigcommerce/catalyst-core:dev:       "node": {
@bigcommerce/catalyst-core:dev:         "__typename": "Product",
@bigcommerce/catalyst-core:dev:         "id": "UHJvZHVjdDo4NTk=",
@bigcommerce/catalyst-core:dev:         "entityId": 859
@bigcommerce/catalyst-core:dev:       }
@bigcommerce/catalyst-core:dev:     },
@bigcommerce/catalyst-core:dev:     "expiryTime": 1729257562395
@bigcommerce/catalyst-core:dev:   },
@bigcommerce/catalyst-core:dev:   {
@bigcommerce/catalyst-core:dev:     "status": "LAUNCHED",
@bigcommerce/catalyst-core:dev:     "expiryTime": 1729256851939
@bigcommerce/catalyst-core:dev:   }
@bigcommerce/catalyst-core:dev: ]







Frustratingly, this actually is lending a lot of credence to the theory that the actual root cause is route.node.id ending in "="

Please see link to previous PR comments with my redirect console logging to confirm, but each one ends in "="

**Pull request merged**